src/Microsoft.NetCore.Analyzers/Core/Security/UseAutoValidateAntiforgeryToken.cs

```diff
@@ -0,0 +1,282 @@
+                                }
+                                else if (actionMethodSymbol.HasAttribute(httpPatchAttributeTypeSymbol))
+                                {
+                                    actionMethodSymbols.Add((actionMethodSymbol, "HttpPatch"));
+                                }
+                            }
+                        }
+                    }
+                }, SymbolKind.NamedType);
+
+                compilationStartAnalysisContext.RegisterCompilationEndAction(
+                (CompilationAnalysisContext compilationAnalysisContext) =>
+                {
+                    if (usingValidateAntiForgeryAttribute && !hasGlobalAntiForgeryFilter && actionMethodSymbols.Count != 0)
+                    {
+                        var visited = new Dictionary<IMethodSymbol, bool>();
+
+                        foreach (var onAuthorizationAsyncMethodSymbol in onAuthorizationAsyncMethodSymbols)
+                        {
+                            CheckIfACertainMethodGetsCalled(
+                                onAuthorizationAsyncMethodSymbol,
+                                visited,
+                                (IMethodSymbol methodSymbol) =>
+                                    (methodSymbol.Name == "ValidateRequestAsync" &&
+                                    (methodSymbol.ContainingType.AllInterfaces.Contains(iAntiforgeryTypeSymbol) ||
+                                    methodSymbol.ContainingType.Equals(iAntiforgeryTypeSymbol))));
+
+                            if (visited[onAuthorizationAsyncMethodSymbol])
+                            {
+                                return;
+                            }
+                        }
+
+                        foreach (var (methodSymbol, attributeName) in actionMethodSymbols)
+                        {
+                            compilationAnalysisContext.ReportDiagnostic(
+                                methodSymbol.CreateDiagnostic(
+                                    Rule,
+                                    methodSymbol.Name,
+                                    attributeName));
+                        }
+                    }
+                });
+
+                // <summary>
+                // Check if this method calls a method, which meets the requirement, directly or indirecly.
+                // </summary>
+                // <param name="methodSymbol">The symbol of the method to be analyzed</param>
+                // <param name="visited">All the methods has been visited and its results</param>
+                // <param name="Requirement">The requirements</param>
+                void CheckIfACertainMethodGetsCalled(IMethodSymbol methodSymbol, Dictionary<IMethodSymbol, bool> visited, RequirementsOfValidateMethod requirements)
+                {
+                    if (!visited.TryGetValue(methodSymbol, out var result))
+                    {
+                        visited[methodSymbol] = false;
+
```

Comment:
I worry about partial results here.  Say you have methods A() and B() and OnAuthorizationAsync(), with calls graphs of:
- A() calls:
  - B()
  - OnAuthorizationAsync()
- B() calls:
  - A()

If we start with CheckIfACertainMethodGetsCalled() on A(), then visited[A()'s method symbol] is gonna start as false.  When looking through A()'s called methods, assuming we see calling B() first, then we'll recurse on CheckIfACertainMethodGetsCalled() on B(), which then recurses again on CheckIfACertainMethodGetsCalled() on A() again, but this time, visited[A()'s method symbol] exists and is false.  visited[B()'s method symbol] will always stays false, even though that's the incorrect result (since B() calls A() which calls OnAuthorizationAsync()).

I don't know if it causes incorrect behavior for this particular analyzer, but it doesn't feel right.

Reply:
Maybe I can change the method description as "Check if there's a method with specific requirements is getting called by another method." Although it didn't get the names of all callers, we only need to know if a certain method is called and bail out early is better. Does this make sense?

Reply:
I'm more concerned with someone copying and pasting this code, and expecting the visited dictionary's results to be correct.

Personally I'd rather it compute correct results, so we don't have to worry about whether or not there are any cases that cause the analyzer to give incorrect diagnostics.